packages/flutter/lib/src/cupertino/dialog.dart

```diff
@@ -410,106 +575,686 @@ class _CupertinoAlertActionSection extends StatelessWidget {
+      bkFillPath.addRect(pressedButtonRects[i]);
+    }
+
+    canvas.drawPath(
+      bkFillPath,
+      _buttonBkPaint,
+    );
+
+    // Create the pressed buttons background path and paint it.
+    final Path pressedBkFillPath = new Path();
+    for (int i = 0; i < pressedButtonRects.length; i += 1) {
+      pressedBkFillPath.addRect(pressedButtonRects[i]);
+    }
+
+    canvas.drawPath(
+      pressedBkFillPath,
+      _pressedButtonBkPaint,
+    );
+
+    // Create the dividers path and paint it.
+    final Path dividersPath = new Path()
+      ..addRect(horizontalDivider)
+      ..addRect(verticalDivider);
+
+    canvas.drawPath(
+      dividersPath,
+      _dividerPaint,
+    );
+  }
+
+  void _drawButtonBackgroundsAndDividersStacked(Canvas canvas, Offset offset) {
+    final Offset dividerOffset = new Offset(0.0, dividerWidth);
+
+    final Path bkFillPath = new Path()
+      ..fillType = PathFillType.evenOdd
+      ..addRect(Rect.largest);
+
+    final Path pressedBkFillPath = new Path();
+
+    final Path dividersPath = new Path();
+
+    Offset accumulatingOffset = offset;
+
+    final List<RenderBox> children = getChildrenAsList();
+    for (int i = 0; i < children.length; i += 1) {
+      final bool isButtonPressed = pressedButtons.contains(i);
+      final bool isPrevButtonPressed = pressedButtons.contains(i - 1);
+      final bool dividerNeeded = i == 0 || !(isButtonPressed || isPrevButtonPressed);
+
+      final Rect dividerRect = new Rect.fromLTWH(
+        accumulatingOffset.dx,
+        accumulatingOffset.dy,
+        size.width,
+        dividerWidth,
+      );
+
+      final Rect buttonBkRect = new Rect.fromLTWH(
+        accumulatingOffset.dx,
+        accumulatingOffset.dy + dividerWidth,
+        size.width,
+        children[i].size.height,
+      );
+
+      // If this button is pressed, then we don't want a white background to be
+      // painted, so we erase this button from the background path.
+      if (isButtonPressed) {
+        bkFillPath.addRect(buttonBkRect);
+        pressedBkFillPath.addRect(buttonBkRect);
+      }
+
+      // If this divider is needed, then we erase the divider area from the
+      // background path, and on top of that we paint a translucent gray to
+      // darken the divider area.
+      if (dividerNeeded) {
+        bkFillPath.addRect(dividerRect);
+        dividersPath.addRect(dividerRect);
+      }
+
+      accumulatingOffset += dividerOffset + new Offset(0.0, children[i].size.height);
+    }
+
+    // Draw all of the button backgrounds.
+    canvas.drawPath(
+      bkFillPath,
+      _buttonBkPaint,
+    );
+
+    // Draw the pressed button backgrounds.
+    canvas.drawPath(
+      pressedBkFillPath,
+      _pressedButtonBkPaint,
+    );
+
+    // Draw all of the dividers.
+    canvas.drawPath(
+      dividersPath,
+      _dividerPaint,
+    );
+  }
+
+  void _drawButtons(PaintingContext context, Offset offset) {
+    for (RenderBox child in getChildrenAsList()) {
```

Comment:
Doesn't defaultPaint do the same thing?

Reply:
I think it does, but since there is other behavior added to the paint() method, I thought it would be more clear to explicitly do the painting.  Otherwise the developer would essentially see:

paintBackground()
paintDividers()
defaultPaint()

and it is not immediately clear to the developer what defaultPaint() is doing with regard to this dialog.

Does that make sense?

Reply:
I meant having _drawButtons just call defaultPaint. I don't have a strong feeling either way. You decide.

Reply:
the advantage of using defaultPaint() is that it is more efficient (doesn't use getChildrenAsList, e.g.), and if we find a way of making it faster, everything benefits all at once.